lib/std/posix.zig

```diff
@@ -4381,51 +4397,108 @@ pub const FStatAtError = FStatError || error{
 /// On WASI, `pathname` should be encoded as valid UTF-8.
 /// On other platforms, `pathname` is an opaque sequence of bytes with no particular encoding.
-/// See also `fstatatZ` and `std.os.fstatat_wasi`.
+/// See also `fstatatZ`.
 pub fn fstatat(dirfd: fd_t, pathname: []const u8, flags: u32) FStatAtError!Stat {
-    if (native_os == .wasi and !builtin.link_libc) {
-        const filestat = try std.os.fstatat_wasi(dirfd, pathname, .{
-            .SYMLINK_FOLLOW = (flags & AT.SYMLINK_NOFOLLOW) == 0,
-        });
-        return Stat.fromFilestat(filestat);
-    } else if (native_os == .windows) {
+    if (native_os == .windows)
         @compileError("fstatat is not yet implemented on Windows");
-    } else {
-        const pathname_c = try toPosixPath(pathname);
-        return fstatatZ(dirfd, &pathname_c, flags);
-    }
-}
-
-/// Same as `fstatat` but `pathname` is null-terminated.
-/// See also `fstatat`.
-pub fn fstatatZ(dirfd: fd_t, pathname: [*:0]const u8, flags: u32) FStatAtError!Stat {
-    if (native_os == .wasi and !builtin.link_libc) {
-        const filestat = try std.os.fstatat_wasi(dirfd, mem.sliceTo(pathname, 0), .{
+    if (native_os == .wasi and !builtin.link_libc)
+        return Stat.fromFilestat(try fstatatWasi(dirfd, pathname, .{
             .SYMLINK_FOLLOW = (flags & AT.SYMLINK_NOFOLLOW) == 0,
-        });
-        return Stat.fromFilestat(filestat);
-    }
+        }));
 
-    const fstatat_sym = if (lfs64_abi) system.fstatat64 else system.fstatat;
-    var stat = mem.zeroes(Stat);
-    switch (errno(fstatat_sym(dirfd, pathname, &stat, flags))) {
+    const pathname_c = try toPosixPath(pathname);
+    return fstatatZ(dirfd, &pathname_c, flags);
+}
+
+/// Same as `fstatat` but for WASI.
+/// `pathname` should be encoded as a valid UTF-8 string.
+pub fn fstatatWasi(dirfd: fd_t, pathname: []const u8, flags: wasi.lookupflags_t) FStatAtError!wasi.filestat_t {
+    var stat: wasi.filestat_t = undefined;
+    return switch (wasi.path_filestat_get(dirfd, flags, pathname.ptr, pathname.len, &stat)) {
         .SUCCESS => return stat,
         .INVAL => unreachable,
         .BADF => unreachable, // Always a race condition.
         .NOMEM => return error.SystemResources,
         .ACCES => return error.AccessDenied,
-        .PERM => return error.AccessDenied,
         .FAULT => unreachable,
         .NAMETOOLONG => return error.NameTooLong,
-        .LOOP => return error.SymLinkLoop,
         .NOENT => return error.FileNotFound,
         .NOTDIR => return error.FileNotFound,
-        .ILSEQ => |err| if (native_os == .wasi)
-            return error.InvalidUtf8
-        else
-            return unexpectedErrno(err),
+        .NOTCAPABLE => return error.AccessDenied,
+        .ILSEQ => return error.InvalidUtf8,
         else => |err| return unexpectedErrno(err),
-    }
+    };
+}
+
+/// Same as `fstatat` except it uses `statx(2)` if the target does not implement `fstatat(2)`.
+fn fstatatLinux(
```

Comment:
My reasoning was that there is a lot of code that calls `std.os.fstat(at)` directly, and having it return a compile error if the `fstatat` syscall isn't implemented won't be a great experience when people upgrade to newer zig versions. I'm fine removing it, but I would like a definitive answer either way (perhaps this can be added to #21738).

Reply:
But does this affect anything other than arc and riscv32? (Support for those architectures is new in this Zig release anyway.)

Reply:
cc @andrewrk given our recent discussion on the future of `std.posix`.